### PR TITLE
Fixes #20248, #20327 - Respect context in CSV export

### PR DESCRIPTION
**Pull request by tbrisker**
Since CSV export is done using an enumerator which is executed by a
different thread to the one that creates it, we need to manually pass in
the context (user, location, organization) to the enumerator.

### CODE REVIEW

File: app/controllers/hosts_controller.rb [hunk 1]
```diff
@@ -60,7 +60,7 @@ def index(title = nil)
         render :index if title && (@title = title)
       end
       format.csv do
-        @hosts = search.includes(included_associations - [:host_statuses, :token, :compute_resource])
+        @hosts = search.preload(included_associations - [:token])
```
Comment (tbrisker) [inline app/controllers/hosts_controller.rb]: not directly related but took advantage of this PR for a small performance improvement.

File: app/helpers/hosts_and_hostgroups_helper.rb [hunk 2]
```diff
@@ -1,12 +1,6 @@
 module HostsAndHostgroupsHelper
   include AncestryHelper
 
-  def model_name(host)
-    name = host.try(:model)
-    name = host.compute_resource.name if host.compute_resource
-    name
-  end
```
Comment (tbrisker) [inline app/helpers/hosts_and_hostgroups_helper.rb:8]: moved this into the host object so that it can be seen in the csv.

File: app/services/csv_exporter.rb [hunk 3]
```diff
@@ -4,13 +4,17 @@ module CsvExporter
...
 
-      columns.map!{|c| c.to_s.split('.').map(&:to_sym)}
-      resources.uncached do
-        resources.reorder(nil).limit(nil).find_each do |obj|
-          csv << CSV.generate_line(columns.map{|c| c.inject(obj, :try)})
+        columns.map!{|c| c.to_s.split('.').map(&:to_sym)}
+        resources.uncached do
+          resources.reorder(nil).limit(nil).find_each do |obj|
+            csv << CSV.generate_line(columns.map{|c| c.inject(obj, :try)})
+          end
```
Comment (tbrisker) [inline app/services/csv_exporter.rb]: Just added spaces with no change due to `User.as` block

File: test/controllers/hosts_controller_test.rb [hunk 4]
```diff
@@ -31,6 +31,18 @@ def host_attributes(host)
     assert_template 'index'
   end
 
+  test "should get csv index with data" do
+    host = FactoryGirl.create(:host, :with_hostgroup, :with_environment, :on_compute_resource)
```
Comment (tbrisker) [inline test/controllers/hosts_controller_test.rb]: I wanted to test the last report field as well but I think it failed due to the way the report factory works.

Comment (tbrisker) [reply test/controllers/hosts_controller_test.rb]: nevermind, figured it out.

File: app/controllers/hosts_controller.rb [hunk 5]
```diff
@@ -60,7 +60,7 @@ def index(title = nil)
         render :index if title && (@title = title)
       end
       format.csv do
-        @hosts = search.includes(included_associations - [:host_statuses, :token, :compute_resource])
```
Comment (ShimShtein) [inline app/controllers/hosts_controller.rb:65]: why do we need `:host_statuses` preloaded?

File: app/services/csv_exporter.rb [hunk 6]
```diff
@@ -4,13 +4,17 @@ module CsvExporter
   def self.export(resources, columns, header = nil)
     header ||= default_header(columns)
     raise ArgumentError, "Columns and header row aren't the same length" unless columns.length == header.length
+    # we need to pass the current user to the enumerator which is executed by a seperate thread
```
Comment (ShimShtein) [inline app/services/csv_exporter.rb]: It took me some time to realize that although `resources` should already be filtered by the user (even though it's in another thread), the actual filter evaluation will be performed in this new thread, so it has to set the user correctly.

Could you please add this info to a comment?

File: test/factories/host_related.rb [hunk 7]
```diff
@@ -174,8 +174,8 @@ def set_nic_attributes(host, attributes, evaluator)
       after(:create) do |host,evaluator|
         evaluator.report_count.times do |i|
           report = FactoryGirl.create(:report, :host => host, :reported_at => (evaluator.report_count - i).minutes.ago)
-          host.last_report = report.reported_at
         end
+        host.update_attribute(:last_report, host.reports.last.reported_at)
```
Comment (ShimShtein) [inline test/factories/host_related.rb:178]: What is the reason for this change?

**Review by ShimShtein — CHANGES_REQUESTED** [review]
It works as expected, a few small comments though.

File: app/controllers/hosts_controller.rb (hunk 5 above)
Comment (tbrisker) [reply app/controllers/hosts_controller.rb:65]: you're right, we don't need it right now.

File: app/services/csv_exporter.rb (hunk 6 above)
Comment (tbrisker) [reply app/services/csv_exporter.rb]: the resources are correctly scoped actually, it's only when loading relations that this fails as there is no current user inside the enumerator.

File: test/factories/host_related.rb (hunk 7 above)
Comment (tbrisker) [reply test/factories/host_related.rb:178]: Otherwise the factory generated host doesn't have the last_report attribute set correctly for some reason. also this removes the n+1 of updating the last report for each report.

**Comment by ShimShtein** [comment]
:+1: ACK

**Comment by tbrisker** [comment]
Looks like the other thread vars (org/loc) should also be passed into the enumerator. setting to WoC for now.

**Comment by tbrisker** [comment]
Updated to also pass location and organization into the enumerator.

File: app/models/concerns/foreman/thread_session.rb [hunk 8]
```diff
@@ -37,6 +37,24 @@ def clear_thread
       end
     end
 
+    # This allows getting and setting all current values in case it's needed,
```
Comment (ShimShtein) [inline app/models/concerns/foreman/thread_session.rb:40]: I can see that there are others that use `Thread.current` to store thread specific context.

May I suggest making the context a bit more generic? It should be pretty easy:
* instead of storing `:user`, `:organization` and `:location` directly in the thread, let's store a new hash there with these keys.
* Modify `*Model` classes to access `Thread.current[:foreman_context][:X]`
* Copying the context from one thread to another would be as simple as deep cloning the context hash.
* Either extend Thread class with `foreman_context` method that will access the new hash, or add this method to `Foreman::ThreadSession` module.

With these changes everyone will be sure that the data is passed to new threads.

**Review by ShimShtein — CHANGES_REQUESTED** [review]
A small suggestion around thread contexts.

File: app/models/concerns/foreman/thread_session.rb (hunk 8 above)
Comment (tbrisker) [reply app/models/concerns/foreman/thread_session.rb:40]: I think this refactoring, while possibly good in the long run, is outside of the scope of this issue, which is needed for 1.15.3 (currently CSV export provides incorrect/partial information). 
I wouldn't want to modify the existing methods here in a minor release, as they are very critical in many places, therefor I only added functionality.

Comment (ShimShtein) [reply app/models/concerns/foreman/thread_session.rb:40]: OK then. Created a refactor [issue](http://projects.theforeman.org/issues/20390) in redmine, just to make it visible.

**Comment by ShimShtein** [comment]
:+1:  Tested on my machine. Filters by user, location and organization.

**Review by dLobatog — APPROVED** [review]
The `compute resource or model` part left me a bit perplexed. My VMs do have a model, but it's definitely not shown in the UI as the `model_name` modified in `hosts_and_hostgroups_helper` shows. That should be fixed at some point though, it's not correct that hosts can only have ONE of model and compute resource.

**Comment by dLobatog** [comment]
Thanks @tbrisker @ShimShtein
